yb-voyager/src/tgtdb/conn_pool.go

```diff
@@ -0,0 +1,114 @@
+package tgtdb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/jackc/pgx/v4"
+	log "github.com/sirupsen/logrus"
+)
+
+var sessionVars = map[string]string{
+	"client_encoding":          "'UTF-8'",
+	"session_replication_role": "replica",
+}
+
+type ConnectionParams struct {
+	NumConnections int
+	ConnUriList    []string
+	SessionVars    map[string]string
+}
+
+type ConnectionPool struct {
+	sync.Mutex
+	params       *ConnectionParams
+	conns        chan *pgx.Conn
+	nextUriIndex int
+}
+
+func NewConnectionPool(params *ConnectionParams) *ConnectionPool {
+	pool := &ConnectionPool{
+		params: params,
+		conns:  make(chan *pgx.Conn, params.NumConnections),
+	}
+	for i := 0; i < params.NumConnections; i++ {
+		pool.conns <- nil
+	}
+	if pool.params.SessionVars == nil {
+		pool.params.SessionVars = map[string]string{}
+	}
+	for k, v := range sessionVars {
+		pool.params.SessionVars[k] = v
+	}
+	return pool
+}
+
+func (pool *ConnectionPool) WithConn(fn func(*pgx.Conn) error) error {
+	var err error
+
+	conn := <-pool.conns
+	if conn == nil {
+		conn, err = pool.createNewConnection()
+		if err != nil {
+			return err
+		}
+	}
+
+	err = fn(conn)
+
+	if err != nil {
+		// On err, drop the connection.
+		conn.Close(context.Background())
+		pool.conns <- nil
+	} else {
+		pool.conns <- conn
+	}
+
+	return err
+}
+
+func (pool *ConnectionPool) createNewConnection() (conn *pgx.Conn, err error) {
+	idx := pool.getNextUriIndex()
+	n := len(pool.params.ConnUriList)
+	for i := 0; i < n; i++ {
+		uri := pool.params.ConnUriList[(idx+i)%n]
+		conn, err = pgx.Connect(context.Background(), uri)
+		if err != nil {
+			log.Warnf("Failed to connect to %q: %s", uri, err)
+			continue
```

Comment:
If we are continuing leaving that server, so i think we should increment this URI function also `getNextUriIndex()`
Because let's say 2 out of 5 target servers are down in that case we will want to distribute equal among the 3 lives ones. So it seems if we don't increment then the live servers(next to dead ones) in your list we will overloaded as compared to others.

Let me know if this makes sense.

Reply:
Changed the implementation so that connections get distributed approximately evenly across all alive nodes.